crypto/evp/evp_lib.c

```diff
@@ -27,6 +28,27 @@
 #include "evp_local.h"
 
 #if !defined(FIPS_MODULE)
+
+EVP_PKEY *OSSL_keygen(OSSL_LIB_CTX *libctx, const char *name, int param)
+{
+    EVP_PKEY *pkey = NULL;
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_from_name(libctx, name, NULL);
+
+    if (ctx != NULL && EVP_PKEY_keygen_init(ctx) > 0) {
+        if (strcmp(name, "RSA") == 0
+                && !EVP_PKEY_CTX_set_rsa_keygen_bits(ctx, param))
+            goto end;
+        if (strcmp(name, "EC") == 0
+                && !EVP_PKEY_CTX_set_ec_paramgen_curve_nid(ctx, param))
```

Comment:
I thoroughly dislike the integer `param`, and even more that we're *again* getting locked in with NIDs.  What happens when we get a key with another sort of parameter?  Are we going to have to have a `OSSL_keygen_ex()` with one more added argument?  And when there's a third?

For the sake of future proofing, an `OSSL_PARAM` array is more future proof...  and using names where possible.  And it means calling `EVP_PKEY_CTX_set_params()` without having to check the EVP_PKEY type.

Reply:
Well, I originally meant the `OSSL_keygen()` function just as a very limited vehicle for concise use in the OpenSSL tests and for implementing RSA- and EC- specific convenience functions, well knowing that that integer parameter was pretty ugly.

Meanwhile due to your remark I found that it can be generalized nicely to
```
 EVP_PKEY *OSSL_keygen(OSSL_LIB_CTX *libctx, const char *name,
                       const char *propq, OSSL_PARAM *params);
```
which is not that convenient to use, but the RSA- and EC- specific convenience functions can be slightly generalized to
```
EVP_PKEY *OSSL_RSA_keygen(OSSL_LIB_CTX *libctx, unsigned int bits);
EVP_PKEY *OSSL_EC_keygen(OSSL_LIB_CTX *libctx, const char *curve);
```
which is still pretty convenient while being sufficient for various uses also within the OpenSSL tests. The EC function now uses the curve name rather than an NID.

I'll push the respective change within the next hours...